### PR TITLE
Allow templates to be located on non-default branch

### PR DESCRIPTION
**Pull request by taras**
## Hey, I just made a Pull Request!

This PR adds the ability to use a template from a non-default branch.

## Motivation

I have a GitHub repository that has a GitHub template. I converted it into a cookiecutter template on a branch. I would like to use this branch with the scaffolder.

## Approach

Checkout to the branch specified by the URL of the template. This solution will be susceptible to #2815 which I'm not addressing here. 

I also changed how the path to the generated output is created. Instead of relying on user-specified component_id, I'm just taking the only generated directory in the template directory. This allows me to generate component_id using cookiecutter variables.  

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [ ] Regression tests added for bug fixes
- [ ] Added a changeset ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))

### CODE REVIEW

File: plugins/scaffolder-backend/src/scaffolder/stages/prepare/github.ts [hunk 1]
```diff
@@ -64,7 +64,22 @@ export class GithubPreparer implements PreparerBase {
         }
       : {};
 
-    await Clone.clone(repositoryCheckoutUrl, tempDir, cloneOptions);
+    const repository = await Clone.clone(
+      repositoryCheckoutUrl,
+      tempDir,
+      cloneOptions,
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/stages/prepare/github.ts:70]: Could we just provide `checkoutBranch` here instead? I imagine the behavior of not switching the branch if it's not found is mainly just confusing, or is there a good reason for it to behave that way?

Comment (taras) [reply plugins/scaffolder-backend/src/scaffolder/stages/prepare/github.ts]: Yes, good call. I never used gitnode before so I'm not familiar with the API. Thanks

**Comment by taras** [comment]
I don't really understand this test failure. 

image

How can runCommand be undefined if it's just being imported from a module? Could Jest mocking be messing up that function? (don't use function mocking usually so I'm not sure how it works)

**Comment by Rugvip** [comment]
@taras Where are you seeing it? Error in CI is something different.

**Comment by taras** [comment]
@Rugvip locally. It's so strange.

**Comment by Rugvip** [comment]
@taras `console.log(runCommand)` :grin:

**Comment by taras** [comment]
@Rugvip it was undefined 😨 but passing on CI... so I'll just go with CI 😄

**Comment by Rugvip** [comment]
@taras Oh yeah it's mocked in `cookiecutter.test.ts`.

I'm guessing the tests are failing for you locally because you have `cookiecutter` installed - A LIL BIT LEAKY TEST 😅

File: plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts [hunk 2]
```diff
@@ -78,8 +78,10 @@ export class CookieCutter implements TemplaterBase {
       });
     }
 
+    const [generated = ''] = await fs.readdir(resultDir);
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts:81]: What's the reason for this change? Could use a clarifying comment why it's safe to assume this.

Also seems like defaulting to `''` is a bit strange, as that would mean the template has no output? Thinking we might wanna just error out in that case.

**Review by Rugvip — COMMENTED** [review]
Sorry, missed this bit the first time round 😅

File: plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts (hunk 2 above)
Comment (taras) [reply plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts:81]: From my PR description,

> I also changed how the path to the generated output is created. Instead of relying on user-specified component_id, I'm just taking the only generated directory in the template directory. This allows me to generate component_id using cookiecutter variables.

The previous implementation relied on `component-id` which was supposed to be passed from the UI. When I created the template, I used cookiecutter variables to define component-id from a name so it takes care sluggifying the component-id. Then I discovered that cookiecutter variables are not available when this service is running. As a result, the Node was throwing an error because it couldn't create a path with an `undefined` segment. 

I couldn't find a way to pull the computed values from cookiecutter. Instead, I took the only generated directory and used it as the path. (this could be incorrect but it seems like it should work reliably).

The test stubs out `runDockerContainer` which means that the result directory is never updated. As a result, `[generated] = await fs.readdir(resultDir)` is undefined. 

There are a bunch of ways to make this better,
1. Figure out a way to allow combining cookiecutter variables and values passed in from the UI
2. Refactor tests to stub out `runDockerContainer` to populate results directory
3. Add tests for `await commandExists('cookiecutter') === true` scenario

Now that I look at it, I should at least do 3. What do you want to do about 1 & 2?

Comment (Rugvip) [reply plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts:81]: Alright, makes sense, please do add a comment explaining what's going on and why. Also likely best to throw an error if `generated` ends up being `undefined`, as there's no way that wouldn't lead to more errors later anyway, better catch that early 😅 

I think `generated` could be `subDir` or smth too?

Regarding the tests, use `mock-fs` for testing file system things. If it would lead to too much refactoring, prolly just `jest.spyOn(fs, 'readdir')`

Comment (taras) [reply plugins/scaffolder-backend/src/scaffolder/stages/templater/cookiecutter.ts]: 2 & 3 are done

**Review by Rugvip — APPROVED** [review]
Thanks! :grin:
